timers: fix processing of nested same delay timers

whitlockjc opened the pull request:
This is a port of https://github.com/nodejs/nodejs/node-v0.x-archive/pull/17203 and https://github.com/nodejs/node-v0.x-archive/pull/25763 which fixes https://github.com/nodejs/node-v0.x-archive/issues/9333, https://github.com/nodejs/node-v0.x-archive/issues/15447, https://github.com/nodejs/node-v0.x-archive/issues/25607 and https://github.com/nodejs/node/issues/5426.

/cc @Fishrock123, @misterdjules, @nodejs/collaborators and @nodejs/tsc

mscdex commented on test/common.js:
```diff
@@ -445,3 +446,11 @@ exports.fileExists = function(pathname) {
   }
 };
+
+exports.busyLoop = function busyLoop(time) {
```
Why add this for just one test?

whitlockjc replied:
Good question.  :)  I'm glad to remove it until others need something similar.  My only worry is removing it means people will probably write their own stuff not knowing that we have a common need for a busy loop.  Your call.

Fishrock123 replied:
Yes, if you could move it to a function in the bottom of the test file that would be better for now.

Fishrock123 replied:
Edit, maybe don't pending https://github.com/nodejs/node/pull/2232

whitlockjc replied:
Just let me know.

Fishrock123 replied:
yeah this should go into the test itself imo

whitlockjc replied:
If this PR is still up for consideration, I'll update this.  I'm not sure where it stands so I defer to your judgement.

Fishrock123 replied:
Seems like there is an actual bug here, so yeas, please update it. :)

whitlockjc replied:
Will do.  Thanks for your help.

mscdex commented:
/cc @bnoordhuis @trevnorris ?

Fishrock123 commented:
Does https://github.com/nodejs/node/pull/2232 need to land first, or does this incorporate that?

Can we just port the test maybe? We should probably preserve julien's commit though.

whitlockjc commented:
@misterdjules should comment.  Last I talked to Julien I was asked to port this over but if there is already work to port things and this is superfluous, no big deal to me.  Just let me know how I can help.

misterdjules commented on test/timers/test-timers-nested.js:
```diff
@@ -0,0 +1,32 @@
+    common.busyLoop(delay);
+
+    process.nextTick(function() {
+      assert.ok(!nestedCalled);
```
I'd suggest adding a comment that this `assert` is done within a `nextTick` callback so that it runs after `A` completed, but before the next turn of libuv's event loop. Otherwise it's not clear for anyone but who wrote the code why it's done this way.

misterdjules commented:
This PR fixes one issue with #2232 by basically refactoring it, so most of the changes in #2232 are superseded by this PR. However, we need `test/parallel/test-timers-blocking-callback.js` and [another one line change](https://github.com/nodejs/node/pull/2232/files#diff-0a5d4868b2b9b17cf9e2c11f1bd1311eR60) from #2232.

whitlockjc commented:
I have made the changes @misterdjules requested and the PR is updated with those changes.  Let me know if there is more to be done and thanks for your help.

misterdjules commented:
@whitlockjc 

>  However, we need test/parallel/test-timers-blocking-callback.js and another one line change from #2232.

I don't see these changes in this PR, do you have some time to update this PR with these changes? Then we'll close github.com/nodejs/node/pull/2232 in favor of this one.

whitlockjc commented:
Sure thing.  I am mobile right now but when I get back to the laptop, I will update this PR with those changes and resolve conflicts.

whitlockjc commented:
I'll take care of this today.  I didn't manage to get laptop time over the holiday.

Fishrock123 commented on lib/timers.js:
```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
+    //
+    // We want to make sure that newly added timer fires in the next turn of the
+    // event loop at the earliest. So even if it's already expired now,
+    // reschedule it to fire later.
```
I'm not sure this is actually a bug. Shouldn't the timer be scheduled in `setImmediate()` if you want to guarantee this?

Also, how does this not only apply to 0ms timeout timers?

> `setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);`

In this example, the 2nd timeout should be caught by https://github.com/nodejs/node/blob/master/lib/timers.js#L68

misterdjules replied:
> I'm not sure this is actually a bug.

It is indeed a grey area. However, this is how node v4.x and later behaves now. It is also how node v0.10.x and v0.12.x used to behave before https://github.com/nodejs/node-v0.x-archive/commit/d38e865fce93c00cb128034202fb1e26987efdbc, which is a regression that is fixed by https://github.com/nodejs/node-v0.x-archive/pull/25763.

> Shouldn't the timer be scheduled in setImmediate() if you want to guarantee this?

Scheduling the timer with setImmediate would basically make it not a timer anymore, as immediates are scheduled and ran in a completely different way.

> Also, how does this not only apply to 0ms timeout timers?

> setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);

The comment in this PR should be clearer and mention that it happens with non-0 delay timers if the callback of the outer timer blocks the JS thread for a duration that is greater or equal to the timer delay. So the snippet should be:

```
setTimeout(function foo() { setTimeout(function bar() {}, 500); codeThatBlocksFor500ms(); }, 500);
```


> In this example, the 2nd timeout should be caught by https://github.com/nodejs/node/blob/master/lib/timers.js#L68

Except that then we'd hit the bug fixed by https://github.com/nodejs/node-v0.x-archive/commit/d38e865fce93c00cb128034202fb1e26987efdbc.

That bug does not have a big impact for timers with short delays, but it can be a problem for timers with large delays that are scheduled before code that blocks for a long time, as it can make the timer fire with a delay of `originalDelay + timeSpentBlockingInOutterTimer`.

Fishrock123 commented on lib/timers.js:
```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
+    // list, because any remaining timer has been added within a callback of a
+    // timer that has already been processed, and thus needs to be processed at
+    // the earliest not in the current tick, but when the rescheduled timer will
+    // expire.
```
This is already how the logic works.

misterdjules replied:
This describes the specific case when a nested timer with the same delay is processed, which is currently (as in, without this change) not handled properly.

Fishrock123 commented on test/timers/test-timers-nested.js:
```diff
@@ -0,0 +1,39 @@
+
+    // Busy loop for the same timeout used for the nested timer to ensure that
+    // we are in fact expiring the nested timer.
+    common.busyLoop(delay);
```
Timeout if this happens is actually the logically correct behavior.

Fishrock123 replied:
Example:

The pattern used here is prone to this sort of error. You should schedule your timeout after you do your work.

A similar thing may happen if, say, you tell an http server to start listening before you attach any handlers to it.

If you schedule a timeout, and it's time passes, and it's in the timeout list, it will timeout as soon as possible. Isn't that expected?

misterdjules replied:
The point is to test that the nested timer's callback is called in the next turn of the event loop, after e.g `process.nextTick` and promises callbacks that have already been scheduled.

Fishrock123 replied:
I don't think my opinion here is valid anymore

Fishrock123 commented:
Sorry, after figuring out how timers work fully for https://github.com/nodejs/node/pull/4007 I'm not convinced this is the correct behavior, or necessary?

whitlockjc commented:
Have we ran `test-timers-blocking-callback.js` and `test-timers-nested.js` as added/changed by this PR and #2232 to see if these tests run successfully and non-flaky?  As for deciding whether this is necessary or not, I defer to @misterdjules as these issues were originally reported by him and initial fixes were by him as well.  I know that when the work was done, there was definitely an issue with timers and this PR reproduced the issue with its tests and fixed the issue with its changes.  In the meantime, I'll go look at #4007 to get acquainted with the new approach.

Fishrock123 commented:
I think the included test isn't actually valid, so it doesn't really matter if it is flaky or not. :s

whitlockjc commented:
It seemed to work with the original sources.  ;)  I guess that doesn't matter now.

Fishrock123 commented:
@whitlockjc I think you may be misunderstanding. It may work but that doesn't mean it is correct. :/

Fishrock123 commented:
Is this PR just trying to make sure the callback is always Async per-se?

whitlockjc commented:
Seeing as I'm not always right and I love to learn, I'd love to hear more about why it's not correct.  At the time the issue was reported, the test was capable of reproducing the reported issue and when the code to fix the issue was in place, the test passed showing the issue had been addressed.  Instead of just saying it's not right, why not provide a little context?  Worst case someone learns something.

Fishrock123 commented:
> Seeing as I'm not always right and I love to learn, I'd love to hear more about why it's not correct.

Sorry, I put some reasoning into https://github.com/nodejs/node/pull/3063#discussion_r46175957 -- basically I'm not sure we guarantee a callback will actually be called at least next tick, and timers by nature timeout as soon as possible.

Fishrock123 commented:
I suppose users may expect it to be async. Would it also be possible to do this by always calling `insert()` within a `setimmediate()` in `active()`? That may be cleaner if it works?

misterdjules commented:
> I suppose users may expect it to be async

It is at least how it behaves now. Having nested timers that expire when their outer timer's callback is done fire asynchronously has the nice side effect of not starving the event loop too.

> Would it also be possible to do this by always calling insert() within a setimmediate() in active()? That may be cleaner if it works?

I don't think we should use setImmediate for rescheduling timers for the reason I mentioned in https://github.com/nodejs/node/pull/3063#discussion_r46333048:

> Scheduling the timer with setImmediate would basically make it not a timer anymore, as immediates are scheduled and ran in a completely different way.

Using `active` with a different timer delay (the remaining time before the timer fires) would create a new underlying timer, and it seems doing that in a general fashion would make the number of underlying timers grow to a large number, for no added benefit except readability. Instead, the code might be made clearer by just factoring it out into separate functions.

Fishrock123 commented on lib/timers.js:
```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
+    // expire.
+    //
+    // See: https://github.com/joyent/node/issues/25607
+    if (now <= first._idleStart) {
```
I don't really understand how this won't be hit when there are just timers scheduled for a later date?

misterdjules replied:
`first._idleStart` is the time at which a timer was scheduled. Since `now` is updated to be the time at the start of `listOnTimeout` execution, `now` is always `> first._idleStart`, _unless_ `first` was scheduled from within the call to `listOnTimeout`.

Fishrock123 replied:
Oh duh, this is `_idleStart` not `_idleTimeout`, I'm sorry.

Fishrock123 commented on lib/timers.js:
```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
+      return;
+    }
+
     if (diff < msecs) {
```
I still don't understand how this won't catch it.

`diff` is guaranteed to be zero or negative here if `now <= first._idleStart` is true. (This might need to be `<=`?)

Is the entire bug that `list.start(msecs - diff, 0);` starts adding if the value os negative?

If so, the logic should be rolled into a negative check in the statement below, I think.

whitlockjc replied:
If you change from `<` to `<=` you could end up processing a timer that was scheduled at the same time we started processing timers which is not what we want.  If you schedule at timer, it should wait at least one tick before being processed.  But this line you're commenting on is unchanged so I'm not sure if agreeing on a change should be fixed here anyways.  Did I misunderstand?

whitlockjc commented:
I have made the changes @Fishrock123 has requested and updated this PR to be merge ready without conflict.  `make && make check` run without issue.

trevnorris commented:
@Fishrock123 did you completely break this :(

Fishrock123 commented:
@trevnorris break this how?

whitlockjc commented:
I will see what changes were made that cause the issue.  As of my last comment yesterday, there were no merge conflicts.  Stand by.

whitlockjc commented:
It appears that way.  60f8c1acf4462ff546fbdda3e071694767b86f1c and 67963c8c6643f0f6f406b0645814faf88f88b679 were committed and they alter `lib/timers.js` quite a bit.  I'll review the changes by @Fishrock123 to see if this PR is even relevant anymore and if it is, I'll fix the PR.

whitlockjc commented:
I'm updating the PR as we speak.  @Fishrock123's changes and mine do not conflict.

whitlockjc commented:
The work is done...*(for now)*.  ;)

jasnell commented:
/cc @nodejs/ctc

jasnell commented:
@nodejs/lts ... any thoughts on whether we should consider this one for LTS?

Fishrock123 commented:
@jasnell Yes, this should get backported.

jasnell commented:
Is this ready to land?

misterdjules commented:
Just a reminder to close (and label with lts labels accordingly) https://github.com/nodejs/node/issues/5426 when this lands. We might want to include that issue in the commit message too.

The current commit message currently mentions an issue from joyent/node-v0.x-archive, I suggest to change that mention to include the full link to the issue, otherwise it's not clear in which repository to look.

onnlucky commented:
I had filed #5864 which this pull request fixes. Good.

But perhaps we can make it slightly more nice. The check if we have found a newly scheduled timer or if this timer was not to fire, is essentially the same, one checks for `diff <= 0` other for `diff < msecs`, while msec is 1 or more. The way to deal with both situations is also the same: delay for `msecs - (TimerWrap.now() - timer._idleStart)` or `0`. This is the part the second test gets wrong it bases delay off the originally captured now, which calculates a delay as if all the callbacks didn't cost any time to process.

Perhaps we can turn the code into, one check, catches both cases, and calculates a correct delay.
```
    // When we find a timer in this list that is not yet ready to fire, delay until it is ready.
    // notice, when diff <= 0 we have found a timer that was scheduled, while processing this list
    // notice also, the diff is to the start of processing this list, but delay is calculated against now
    // and notice, TimerWrap.now() returns uv_now() *after* calling uv_update_time()
    if (diff < msecs) {
      var delay = msecs - (TimerWrap.now() - timer._idleStart);
      if (delay < 0) delay = 0;
      this.start(delay, 0);
      debug('%d list wait for %d because diff is %d', msecs, delay, diff);
      return;
    }
```

whitlockjc commented:
@misterdjules I will update the PR to link to #5426.

whitlockjc commented:
@onnlucky I find it easier to understand the two different circumstances to handle them separately but that is just my opinion on the matter.

onnlucky commented:
@whitlockjc You are right, but that second case needs fixing, callbacks do cost time, any new delay must take that into account. But the very first comment says: Why this way? Performance. Doing two checks what could have been the same check, is not in line with that goal, imho.

whitlockjc commented:
The goal of this commit is to fix a bug where a timer could be fired on the same tick that it was created.  Any enhancements or performance fixes outside of this should be a separate commit.

whitlockjc commented:
@misterdjules Done.

onnlucky commented:
The code already took care of the nested timers case, it merely miscalculates the delay when priming the timer. See diff goes negative, so instead of shortening msec, it would subtract a negative, lengthening it. And it uses the wrong time to subtract from, lengthening it more.

The fix is to correctly calculate the delay.

Even if you don't agree, can we not enlarge the scope of an issue when review finds more related things to fix? Or must we add more process to such things? (Not to put you on the spot, I am perfectly happy either way, just asking for the way to go.)

whitlockjc commented:
I don't disagree that it could be better, I'm just not wanting to mix a bug fix commit with an enhancement.

Fishrock123 commented:
Eh, LGTM i suppose. CI: https://ci.nodejs.org/job/node-test-pull-request/2150/

I'll try to address anything else after it lands.

Fishrock123 commented on lib/timers.js:
```diff
@@ -166,6 +166,37 @@ function listOnTimeout() {
+    // the earliest not in the current tick, but when the rescheduled timer will
+    // expire.
+    //
+    // See: https://github.com/joyent/node/issues/25607
```
nodejs/node-v0.x-archive#25607, please use the updated link `https://github.com/nodejs/node-v0.x-archive/issues/25607`

TheAlphaNerd commented:
@Fishrock123 thoughts on this landing?

whitlockjc commented:
@misterdjules' comment [here](https://github.com/nodejs/node/pull/3063#issuecomment-160186902) suggests there is still more work to make this PR ready.  Anyone else have any other needs/wants that do not mix enhancement with bug fixes?  If not, I'll get this taken care of and request a re-review.

zhangzifa commented:
@whitlockjc 

With this fix, please consider the following test case:

```js
var sleepMs = function (ms) {
  var startTime = new Date().getTime();
  while (new Date().getTime() < startTime + ms);
};
setInterval(function() {
  console.log(new Date()); 
  sleepMs(2000);
}, 5000);
```
The output

```js
Wed Jul 06 2016 08:42:37 GMT+0800 (CST)
Wed Jul 06 2016 08:42:44 GMT+0800 (CST)
Wed Jul 06 2016 08:42:51 GMT+0800 (CST)
Wed Jul 06 2016 08:42:58 GMT+0800 (CST)
Wed Jul 06 2016 08:43:05 GMT+0800 (CST)
Wed Jul 06 2016 08:43:12 GMT+0800 (CST)
Wed Jul 06 2016 08:43:19 GMT+0800 (CST)
...
```
As the callback can be finished within `repeat` (5000ms here), in fact, it takes 2000ms, the interval between two timeouts should be `repeat`, not 5000+2000, what's your opinion?

It's a fact that js timer is not accurate, however, we should make it as accurate as possible. 

This is important for some applications, especially for the IoT application.

misterdjules commented:
@zhangzifa The behavior that your [latest comment](https://github.com/nodejs/node/pull/3063#issuecomment-230643962) illustrates is different than the problem that this PR fixes.

This PR is about making nested timers scheduled with `setTimeout` have their callback called at a time that is not offset by the execution of the outer timer's callback too much. In other words, in the following case:

```
setTimeout(function foo() {
  setTimeout(function bar() {
  }, 1000);
  
  // some code that runs on CPU
}, 1000);
```

currently node versions 4.x and later will call the function `bar` at a time that is equivalent to the time when the function `foo` is called + __2x__ the time it takes for the code that runs on CPU to complete, instead of calling it at a time equivalent to the time when `foo` is called + __1x__ the time it takes for the code that runs on the CPU to complete.

That problem doesn't break the (loose) guarantees that node gives about when a timer's callback runs, but we can fix it and provide a behavior that seems to be more reasonable.

The behavior that your latest comment describes is that nested timers scheduled with `setInterval` should fire at a time that does not depend _at all_ on the execution time of the `setInterval` callback, which is a _different_ problem being discussed in #7346.

I would thus suggest that you comment in #7346 and not in this PR.

Thank you!

zhangzifa commented:
@misterdjules 

I agree with you that this PR changes **+ 2x** to **+ 1x**, which make 4.x 5.x 6.x the same behavior as 0.10.x and 0.12.x.

My opinion is it should be **+ 0x**. 

I have raised a new issue [#7554](https://github.com/nodejs/node/issues/7554).

The fix listed [#7554](https://github.com/nodejs/node/issues/7554) includes this #3063 already. And it works for both setTimeout and setInterval.

whitlockjc commented:
I agree it should be ~`5000` and not ~`5000+2000`.  Since the internal sleep is less than the timer, it should complete prior to the timer completing.  I'd have to see why it's happening to know if it should be fixed with this PR.  timers are known to have other issues, some of which were brought up in the thread above, and I personally do no think it's ideal to mix the existing PR with other bug fixes or enhancements.

zhangzifa commented:
Ok, hope this PR merged soon.

onnlucky commented:
Hi guys. I would like to point out, again, that whenever you calculate the delay for the next group of timers, you have to look at the current time, not a time captured from before processing callbacks.
```
this.start(..., 0);
```
that statement is a bug if on the dots you don't take care of the current time.

Fixing that bug, and the solution to nested timers and the cpu heavy timers and the nested timers in cpu heavy timers all come down to the same thing.

But since this pr takes care of it correctly, we can merge it and fix half the timer issues.

Alternatively, we can meld this with line https://github.com/whitlockjc/node/blob/0511688afa7ea8f5cd6d6e7aaa624650eef1d929/lib/timers.js#L202 and fix both.

Fishrock123 commented:
@onnlucky that's what I've been trying to point out...

Fishrock123 commented:
@onnlucky Turns out making a test case for that is hard though, so I guess no-one believes me. 
¯\\\_(ツ)\_/¯

whitlockjc commented:
I don't disagree but like @misterdjules points out, that is another bug and not specific to this one and should be addressed in its own bug.  Mixing new enhancement/bug fixes in an already established PR for another bug does not seem ideal.  _(Maybe you're not suggesting this but just in case...)_

Is there an issue for the issue where timers are basing their delay off the wrong thing?

whitlockjc commented:
As for testing, we could just get the time in between timer fires for a `setInterval` and check for the deviation.  @zhangzifa gave a good code sample that reproduced it but his example used logging to print the deltas where I think we could use the test to check the delta values.

whitlockjc commented:
I'll have the PR updated for @misterdjules requested changes from #2232 shortly.

whitlockjc commented:
PR has been updated.  One thing worth mentioning is that the original PR modified `test/common.js` to add a `busyLoop` function.  This was asked to be removed and put locally into `test/timers/test-timers-nested.js`.  Since `test/parallel/test-timers-blocking-callback.js` uses the `busyLoop` as well, `test/common.js` was updated to have the `busyLoop` function once more and `test/timers/test-timers-nested.js` was updated to use `common#busyLoop` again.

Fishrock123 commented:
As a basic review: Links are still not updated. Also, `timers/` tests are not run, please use `parallel/`.

---

As a secondary issue, the statement should **still** go below into the block below. Not only does it also pass the tests, but it is a lot cleaner, and should fix the other issue. (The other issue, which has now had more reports than this issue.)

Here is a basic diff of what I'm talking about:

```diff
diff --git a/lib/timers.js b/lib/timers.js
index 9dbae32..2d520f3 100644
--- a/lib/timers.js
+++ b/lib/timers.js
@@ -169,7 +169,11 @@ function listOnTimeout() {
     // Check if this loop iteration is too early for the next timer.
     // This happens if there are more timers scheduled for later in the list.
     if (diff < msecs) {
-      this.start(msecs - diff, 0);
+      var timeRemaining = msecs - (TimerWrap.now() - timer._idleStart);
+      if (timeRemaining < 0) {
+        timeRemaining = 0;
+      }
+      this.start(timeRemaining, 0);
       debug('%d list wait because diff is %d', msecs, diff);
       return;
     }
```

whitlockjc commented:
Thanks @Fishrock123, I'll update the PR per your feedback.  I'm not sure what links you're referring to because the `Fixes` and `PR-URL` links are all valid.

Fishrock123 commented:
@whitlockjc Any `joyent/node` links in the code comments.

whitlockjc commented:
Ah, that makes sense.  Thanks.  :)

Fishrock123 commented:
(Or, at least if there is a reason it shouldn't be there the current tests do not account for it.)

Fishrock123 commented on lib/timers.js:
```diff
@@ -160,16 +160,51 @@ function listOnTimeout() {
+  debug('now: %d', now);
 
-  var diff, timer;
+  var diff, timer, timeRemaining;
```
Please just define it inline, we've been moving away from these old "optimizations" -- which really don't matter in todays optimizing JITs

whitlockjc replied:
But there are two places that use the same variable.  I guess that might not be accurate now that you're suggesting the first block isn't necessary.

Fishrock123 commented on lib/timers.js:
```diff
@@ -160,16 +160,51 @@ function listOnTimeout() {
+      debug(msecs + ' list wait because timer was added from another timer');
+      this.start(timeRemaining, 0);
+      return;
+    }
```
@whitlockjc This entire block is now unnecessary -- remove it and run the tests. :)

whitlockjc replied:
Yes...I see what we did there.  I'm an idiot.

Fishrock123 replied:
Nah, timers code is just very involved, for all of us. :P

Fishrock123 commented on test/parallel/test-timers-blocking-callback.js:
```diff
@@ -0,0 +1,80 @@
+
+/*
+ * This is a regression test for https://github.com/joyent/node/issues/15447
+ * and https://github.com/joyent/node/issues/9333.
```
more links still

whitlockjc replied:
I didn't think to fix links in other people's code. :P

whitlockjc commented:
PR updated per @Fishrock123's feedback.  Sorry for the extra iteration, multitasking has failed me again.

whitlockjc commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3203/

Fishrock123 commented on lib/timers.js:
```diff
@@ -169,7 +169,11 @@ function listOnTimeout() {
     // This happens if there are more timers scheduled for later in the list.
     if (diff < msecs) {
-      this.start(msecs - diff, 0);
+      var timeRemaining = msecs - (TimerWrap.now() - timer._idleStart);
```
@whitlockjc Did you still want to put some of them comments here?

I'd prefer if we tried to make them a bit more concise.. but if that's not possible, it's not a huge deal

whitlockjc replied:
I thought about it but since this is now that the code fixing this issue is not on its own, it didn't seem like a decent place to do it.  But we could mention the issue(s) where this block is addressing bugs.

whitlockjc replied:
I've thought about this a little more and now that the current PR code just makes timers work properly without any explicit handling for specific bugs, it would be hard to throw in extra comments without it becoming confusing _(in my opinion)_.  I defer to your judgement but that's my opinion.

Fishrock123 replied:
A little comment on what is going on here would be preferred then

whitlockjc commented:
CI looks good.

whitlockjc commented:
Pinging @Fishrock123 and @misterdjules to take another peek at this.  CI is passing and I think all feedback has been addressed.  I'm ready to address more feedback if/when it comes but we should be ready, or close to ready, to merge this.

misterdjules commented on test/parallel/test-timers-blocking-callback.js:
```diff
@@ -0,0 +1,80 @@
+  ++nbBlockingCallbackCalls;
+
+  if (nbBlockingCallbackCalls > 1) {
+    latestDelay = Date.now() - timeCallbackScheduled;
```
This test should use `process.binding('timer_wrap').Timer.now()` instead of `Date.now()`. I'm not sure why the same test in v0.10 and v0.12 use `Date.now()`, I suspect this is something we overlooked when writing/reviewing that code.

whitlockjc replied:
It's my fault, I should had reviewed the new test being added from #2232.

misterdjules commented on test/parallel/test-timers-blocking-callback.js:
```diff
@@ -0,0 +1,80 @@
+    // block by busy-looping to trigger the issue
+    common.busyLoop(TIMEOUT);
+
+    timeCallbackScheduled = Date.now();
```
Same as [above](https://github.com/nodejs/node/pull/3063/files#r70296792).

misterdjules commented:
@whitlockjc I have only two more comments, and I think I'll be ready to approve this change. Thank you very much for your patience, I think we're very close to merging this :)

Fishrock123 commented:
Commit message needs updating. LGTM with nits fixed

Fishrock123 commented:
Also make sure you link to the original PR and maybe say it is a port of sorts from that commit

whitlockjc commented:
@Fishrock123 What are your suggestions for the commit description/message changes?  Mentioning the original PR is already done in the commit.  I can do the same for `test/parallel/test-timers-blocking-callback.js` if you want as well.

whitlockjc commented:
I've updated the PR with @misterdjules suggested changes.  I also updated the commit message to include the two issues being fixed and made sure the list of `Fixes` are complete _(within reason because there are a number of mentions to this PR)_.

misterdjules commented:
@whitlockjc Thanks for the updates! Do you mind going through another CI run?

whitlockjc commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3297/

whitlockjc commented:
Build failure seems unrelated to this and more related to Jenkins itself.  I'll let you review the output yourself but I think we _should_ be ready.

misterdjules commented:
Build failures look unrelated to me too. LGTM, thank you for everyone's hard work and patience!
I would suggest waiting for @Fishrock123's approval too before landing this change.

whitlockjc commented:
Works for me.  When @Fishrock123 gives the _LGTM_, I'll land if he doesn't.  Thanks for everyone's help.

Fishrock123 commented:
LGTM if you re-look for `https://github.com/joyent/node` links and change them to `nodejs/node-v0.x-archive`

whitlockjc commented:
I fixed the remaining `joyent/node` URIs.

whitlockjc commented:
Landed in 5aac4c4.

misterdjules commented:
@whitlockjc Awesome, thank you again! Next time, please remember to add `Reviewed-By:` lines for each reviewer in the commit message before you push it. It's OK to miss that once in a while though, so please don't force push an updated commit message.

whitlockjc commented:
Crap, I forgot that part.  My bad.  I really did read through the docs, I just must had missed that line.  :(

misterdjules commented:
@whitlockjc It's really not a big deal :) I personally use [git-apply-pr](https://www.npmjs.com/package/git-apply-pr) to land PRs. It automatically adds most of the metadata for me, but I still check that commit message conform to the collaborators guide before pushing. It doesn't prevent all problems, but it definitely helps!

whitlockjc commented:
I'm installing that now as well.  Thanks a lot.

bnoordhuis commented:
For the record: #7866 is caused by this PR although it it may not be a regression in a strict sense.

whitlockjc commented:
I'll take a peek.

misterdjules commented:
@bnoordhuis Thanks for the heads up, I responded in #7866.

TheAlphaNerd commented:
Not 100% if this should land in `v4.x`. Thoughts @misterdjules / @whitlockjc / @bnoordhuis 

It seems like there is a non zero chance this could break production code, even if the production code itself is wrong.

As `v6` is nearing LTS and this exists in there I would opt to simply pass on this for `v4` unless one of you says otherwise.

Adding `dont-land`, please feel free to change

Fishrock123 commented:
@TheAlphaNerd This should be backported along with any fixes.

I'm not sure how it could break, but the new behavior is definitely the expected one. (And inferred by docs too)

Fishrock123 commented:
Here's another issue about this problem: https://github.com/nodejs/node/issues/8354#issuecomment-243782863

@TheAlphaNerd lmk if you need any help backporting

TheAlphaNerd commented:
@nodejs/lts how does everyone feel about backporting this given the above information

@Fishrock123 is this commit dependent on your other timers changes?

Fishrock123 commented:
I don't _think_ so?

TheAlphaNerd commented:
@Fishrock123 will need help with the backport if you can
